core/embed/trezorhal/unix/touch/touch.c

```diff
@@ -51,65 +179,18 @@ uint32_t touch_read(void) {
   int ev_type = 0;
 
   while (SDL_PollEvent(&event) > 0) {
-    switch (event.type) {
-      case SDL_MOUSEBUTTONDOWN:
-        if (is_inside_display(event.button.x, event.button.y)) {
-          ev_x = event.button.x - sdl_touch_offset_x;
-          ev_y = event.button.y - sdl_touch_offset_y;
-          ev_type = TOUCH_START;
-        }
-
-        break;
-
-      case SDL_MOUSEBUTTONUP:
-        if (_touch_detected) {
-          if (is_inside_display(event.button.x, event.button.y)) {
-            ev_x = event.button.x - sdl_touch_offset_x;
-            ev_y = event.button.y - sdl_touch_offset_y;
-          } else {
-            // use last valid coordinates
-            ev_x = _touch_x;
-            ev_y = _touch_y;
-          }
-          ev_type = TOUCH_END;
-        }
-        break;
-
-      case SDL_MOUSEMOTION:
-        if (is_inside_display(event.motion.x, event.motion.y)) {
-          ev_x = event.motion.x - sdl_touch_offset_x;
-          ev_y = event.motion.y - sdl_touch_offset_y;
-
-          if (_touch_detected) {
-            ev_type = TOUCH_MOVE;
-          } else if (event.motion.state != 0) {  // button pressed ?
-            // simulate TOUCH_START
-            ev_type = TOUCH_START;
-          }
-        } else {
-          if (_touch_detected) {
-            // use last valid coordinates and simulate TOUCH_END
-            ev_x = _touch_x;
-            ev_y = _touch_y;
-            ev_type = TOUCH_END;
-          }
-        }
-        break;
+    if (input_state == IDLE || input_state == MOUSE_ACTIVE ||
```

Comment:
What will happen if you follow this sequence?

1) Press the mouse button.
2) Press the left key.
3) Release the mouse button.
4) Release the left key.

Does this sequence lead to consistent behavior?

Reply:
The left key was registered. Now the key events are ignored if they're coming from pressed button (determined by `repeat`).
https://github.com/trezor/trezor-firmware/pull/3619/commits/f248d6218433721716ffe2372555eb65fdeb0fc2

Reply:
I still don't think it will behave absolutely correctly if you combine mouse and buttons or multiple buttons together. However, I also don't think it's worth further attention since it's just an emulator.

Reply:
I believe now it should behave correctly
https://github.com/trezor/trezor-firmware/pull/3619/commits/bedeb8a1be33d6fbc7c7ecb04adbfcab8514bb98